wardrobe/timestamp/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+        // We don't expect any additional peeks typically, but they are harmless.
+        ensure!(!peek_data.is_empty(), Self::Error::MissingPreviousTimestamp);
+        let old_timestamp = peek_data[0]
+            .payload
+            .extract::<Timestamp>()
+            .map_err(|_| Self::Error::BadlyTyped)?;
+
+        // Compare the new timestamp to the previous timestamp
+        ensure!(
+            new_timestamp.time >= old_timestamp.time + T::MINIMUM_TIME_INTERVAL,
+            Self::Error::TimestampTooOld
+        );
+
+        // Make sure the block height from the previous timestamp matches the previous block height.
+        ensure!(
+            new_timestamp.block == old_timestamp.block + 1,
+            Self::Error::PreviousTimestampWrongHeight,
+        );
+
+        Ok(0)
+    }
+
+    fn is_inherent(&self) -> bool {
+        true
+    }
+}
+
+impl<V: Verifier + From<UpForGrabs>, T: TimestampConfig + 'static> TuxedoInherent<V, Self>
+    for SetTimestamp<T>
+{
+    type Error = sp_timestamp::InherentError;
+    const INHERENT_IDENTIFIER: sp_inherents::InherentIdentifier = sp_timestamp::INHERENT_IDENTIFIER;
+
+    fn create_inherent(
+        authoring_inherent_data: &InherentData,
+        previous_inherent: Option<(Transaction<V, Self>, H256)>,
+    ) -> tuxedo_core::types::Transaction<V, Self> {
+        // Extract the current timestamp from the inherent data
+        let timestamp_millis: u64 = authoring_inherent_data
+            .get_data(&sp_timestamp::INHERENT_IDENTIFIER)
+            .expect("Inherent data should decode properly")
+            .expect("Timestamp inherent data should be present.");
+        let new_timestamp = Timestamp {
+            time: timestamp_millis,
+            block: T::block_height(),
+        };
+
+        log::debug!(
+            target: LOG_TARGET,
+            "🕰️🖴 Local timestamp while creating inherent i:: {timestamp_millis}"
+        );
+
+        let mut peeks = Vec::new();
+        match (previous_inherent, T::block_height()) {
+            (None, 1) => {
+                // This is the first block hack case.
+                // We don't need any inputs, so just do nothing.
+            }
+            (None, _) => panic!("Attemping to construct timestamp inherent with no previous inherent (and not block 1)."),
+            (Some((_previous_inherent, previous_id)), _) => {
+                // This is the the normal case. We create a full previous to peek at.
+
+                // We are given the entire previous inherent in case we need data from it or need to scrape the outputs.
+                // But out transactions are simple enough that we know we just need the one and only output.
+                peeks.push(OutputRef {
+                    tx_hash: previous_id,
+                    // There is always 1 output, so we know right where to find it.
+                    index: 0,
+                });
+            }
+        }
```

Comment:
This whole block is the consequence of this tuple matching decision. Not my favorite.

Reply:
It will be completely removed once we can put the extrinsic in genesis block.
@JoshOrndorff can you remind me why it would be a problem if we put the timestamp in the genesis block as raw data rather than an extrinsic?

Reply:
First regarding the code style and readability. I agree it isn't the easiest thing in the world to read. It would actually look a lot more idiomatic if I removed all the comments, but I thought they helped explain the flow. An alternative is to use `if` `else if` `else` but we definitely need these three branches at least until we have a genesis timestamp.

Now regarding the genesis timestamp. My main motivation for putting a timestamp extrinsic in the genesis block is so that we don't need a the first-block-hack in the inherent logic. It isn't because I really want the genesis block to be timestamped. So the genesis timestamp needs to be where the inherent logic expects it, as an extrinsic.

And as a followup point, where else would you put it? The only other place to jam arbitrary data is in the header's digest field.

Reply:
> where else would you put it? 

I might not be right, but we initialize storage with UTXOs that contain money, could we do the same with a timestamp? 
However I think the value would be hard-coded so I'm not sure it would work well in real use cases - you should know before compiling the timestamp of when you will run the chain for the first time.

Reply:
Oh, I see what you mean. Like put the timestamp UTXO in the genesis storage but no actual timestamp extrinsic in the genesis block.

I don't think that would fully solve the problem. It's true that there would be a timestamp there to peek at in block 1. That is good.

But now consider you are the node that needs to author block 1. How do you know what OutputRef to peek at in your timestamp inherent? The usual way is to scrape the previous block to find it. But if there is no transaction in the previous (genesis) block there is nothing to scrape.